### ISSUE TITLE
ROHF conventions

### ISSUE THREAD

**Issue by berquist** [cclib/cclib#301]
Reading https://github.com/cclib/cclib/pull/298, I realized there aren't any conventions that I know of for what cclib does in the case of ROHF calculations. Is this true?

These are results for DVB cation, ROHF rather than UHF:

```
Attempting to read dalton.out
homos:
[34]
Attempting to read gamess.out
homos:
[34 33]
Attempting to read gaussian.out
homos:
[34]
Attempting to read molpro.out
homos:
[34 33]
Attempting to read nwchem.out
homos:
[34]
Attempting to read orca.out
homos:
[34]
Attempting to read qchem.out
homos:
[34 33]
```

No Psi4 result because it doesn't handle UHF/ROHF right now, or the latest version of the program; I'm making major changes to the parser and test files.

I think we should have unit tests for restricted open-shell calculations, which should be defined as (to start with):

```
assert len(data.homos) == 2
assert len(data.mocoeffs) == 1
assert len(data.moenergies) == 1
```

**Comment by langner** [comment]
Very good idea. Either unit tests or regressions will work.

**Comment by bwang2453** [comment]
A related question: how do we put ROHF into metadata? Right now, we only have one boolean variable called `unrestricted`.

**Comment by berquist** [comment]
Since https://github.com/cclib/cclib/pull/298 will bring us the property `closed_shell`, my thought is that there could be another property, `spin_restricted`.

For a closed-shell system, `spin_restricted = True` would be RHF and `spin_restricted = False` would be RHF treated as UHF.

For an open-shell system, `spin_restricted = True` would be ROHF and `spin_restricted = False` would be UHF.

The special case of an open-shell singlet would be `spin_restricted = False` and `closed_shell = True`, so just like a doublet, triplet, etc. but the HOMO indices for both alpha and beta would be the same.

**Comment by berquist** [comment]
Related: [for systems with no beta electrons](https://github.com/cclib/cclib-data/blob/master/NWChem/NWChem6.0/hydrogen_atom_UHF_cc-pVDZ.out), for both UHF and ROHF, `homos` should be `[0 -1]`, not `[0 0]` which is an open-shell singlet. It would be up to people running no-beta-electron calculations to recognize that this index is invalid :)

**Comment by langner** [comment]
@berquist Could you explain that (`homos`being `[0 -1]`) in more detail? I don't get it.

**Comment by berquist** [comment]
Take 3 electrons in 2 orbitals, with one more alpha electron than beta. The alpha HOMO is 2, the beta HOMO is 1, so converting to zero-based indexing gives `homos == [1 0]`.

Now remove the beta electron. There are no more occupied beta-spin orbitals, so choose the beta "HOMO" to be 0. This makes more sense if you consider that number to be the electron count. Converting that to zero-based indexing gives `homos == [1 -1]`.

Those descriptions go with the first two diagrams (top row):

<img src="https://cloud.githubusercontent.com/assets/727571/17740073/407d739a-6465-11e6-8846-dc0188cf1e89.png" width="500" />

The second two diagrams (bottom row) correspond to the hydrogen atom, and then a molecule with no electrons.

So, I think it makes more sense from a convention point of view, and the zero-based indexing makes it look a little funny.

**Comment by berquist** [comment]
For reference, a PR containing some of these updates is https://github.com/cclib/cclib/pull/304.

**Comment by langner** [comment]
Do we use this kind of zero based indexing in cclib somewhere?

**Comment by ATenderholt** [comment]
```mocoeffs``` comes to mind. If you wanted the MO coeffs for the alpha HOMO, you'd simply do ```mocoeffs[homos[0]]```. Of course, this breaks with the edge case where ```homos[0] == -1```.

**Comment by langner** [comment]
I'm starting to understand. I'm not sure if `-1` is the way to go or maybe something like `None` or `nan`. Lack of beta electrons means there is no HOMO in the end.

**Comment by berquist** [comment]
Yes, the point is for the index to be meaningless, but `-1` is a little too dangerous for my taste, and I had forgotten about `nan` and `None`. My choice of the three would be `None`; it reflects the fact that there is no HOMO rather than a nonsensical value, and can be checked in a more Pythonic way:

```
In [8]: bool(nan)
Out[8]: True

In [9]: bool(None)
Out[9]: False

In [10]: bool(-1)
Out[10]: True
```

**Comment by berquist** [comment]
`numpy.nan` and `None` will not work as long as `data.homos` is an integer NumPy array. Aside from changing code that checks for the shape (rather than length) of the array, would there be a problem or any objections with changing `data.homos` to a list of integers?

**Comment by langner** [comment]
Why can't we use `numpy.nan` in a NumPy array?

**Comment by berquist** [comment]
Even though it's "not a number", `numpy.nan` behaves like a float that can't be typecast.

```
In [1]: import numpy as np

In [2]: a = np.ones(3, dtype=np.int32)

In [3]: a
Out[3]: array([1, 1, 1], dtype=int32)

In [4]: a[2] = np.nan
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-02432512a9c3> in <module>()
----> 1 a[2] = np.nan

ValueError: cannot convert float NaN to integer

In [5]: b = np.ones(3, dtype=int)

In [6]: b[1] = np.nan
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-a7030168cda0> in <module>()
----> 1 b[1] = np.nan

ValueError: cannot convert float NaN to integer
```
I am not a fan of the solution presented here with masked arrays, since it seems overly complex: http://stackoverflow.com/questions/12708807/numpy-integer-nan

**Comment by langner** [comment]
I think Eric's previous proposal to use a list of integers is the best one.

**Comment by langner** [comment]
Thinking about this a bit more, changing the type of `homos` is somewhat unfortunate since it introduces a backwards incompatibility, which we have successfully avoided over the years.

Would another option be to reduce the length of the `homos` array when there are no beta electrons?

**Comment by berquist** [comment]
Yes, you could derive the fact that there are no beta electrons by checking for either of the UHF or ROHF conditions given in https://github.com/cclib/cclib/issues/301#issuecomment-239636489, then `data.homos.shape == (1,)`.

Similarly, for the case of no electrons at all:

```python
In [6]: c = np.array([], dtype=int)

In [7]: c.shape
Out[7]: (0,)
```

**Comment by berquist** [comment]
> Would another option be to reduce the length of the `homos` array when there are no beta electrons?

Turns out this isn't going to work either. It is possible to have beta electrons only.

Can we have attributes `nalpha` and `nbeta`?

**Comment by langner** [comment]
> Can we have attributes nalpha and nbeta?
I like this idea, because that would be explicit. Do you think there is some natural way to combine those into one attribute?

**Comment by berquist** [comment]
Let's kill a bunch of birds with one stone. I'm out so I can't have nice
text formatting, but here is what I propose.

nalpha and nbeta can be combined into an integer numpy array called
nelectrons that will always be shape (2,) no matter what.

The number of electrons in a system is completely unambiguous when compared
to homos, which is problematic with symmetry-adapted orbitals or MCSCF.

If we by default derive homos from the nelectrons array, we can later
customize homos to be something else if necessary. There may be code for
occupations in symmetry irreparable somewhere, but that's a separate
project.

Another advantage of deriving homos from nelectrons would be potential
elimination of lots of code that tries to figure out homos from orbital
indices.

**Comment by langner** [comment]
Closed-reopened by msitake. Pushing back for now.

**Comment by berquist** [comment]
Rereading this and it's all over the place. What is the best way forward here? It looks like we might have agreed on having `nelectrons = [nalpha, nbeta]`, but it still isn't clear what `homos` should be for these weird cases. Also, in https://github.com/cclib/cclib/pull/619, the number of electrons is derived from the HOMO indices, when it should be changed to be the other way around.

For reference, with the current master, here are `homos` for a hydrogen atom.

| filename                       | `homos`                        |
|--------------------------------|--------------------------------|
| `dalton_anion_rhf_cc-pVTZ.out` | `array([0], dtype=int32)`      |
| `gamess_rohf_cc-pVTZ.out`      | `array([ 0, -1], dtype=int32)` |
| `gamess_rohf_STO-3G.out`       | `array([ 0, -1], dtype=int32)` |
| `gamess_uhf_cc-pVTZ.out`       | `array([ 0, -1], dtype=int32)` |
| `gamess_uhf_STO-3G.out`        | `array([ 0, -1], dtype=int32)` |
| `nwchem_rohf_cc-pVTZ.out`      | `array([ 0, -1], dtype=int32)` |
| `nwchem_uhf_cc-pVTZ.out`       | `array([ 0, -1], dtype=int32)` |
| `orca_uhf_cc-pVTZ.out`         | `array([ 0, -1], dtype=int32)` |
| `orca_uhf_STO-3G.out`          | `array([ 0, -1], dtype=int32)` |

and I have a load of other similar files that fail to parse; I'll add them as regressions once we decide on a course of action.

**Comment by berquist** [comment]
What do we want to do here?

**Comment by langner** [comment]
Just reread this. I think we wanted to add the `nelectrons` attribute, but that there were also other changes that would be good-to-have (derive `homos` from `nelectrons` and not the other way around).

@berquist does that sound right to you, or was there something that needs to be done that would be backwards incompatible?

**Comment by berquist** [comment]
https://github.com/cclib/cclib/issues/836 increases the importance of this issue.

For a refresher, an ROHF calculation must follow:
```python
assert len(data.homos) == 2
assert len(data.mocoeffs) == 1
assert len(data.moenergies) == 1
```
and
> Just reread this. I think we wanted to add the nelectrons attribute, but that there were also other changes that would be good-to-have (derive homos from nelectrons and not the other way around).

Yes. There is still ambiguity for `homos` in the rare case when you don't have electrons, but by having that as a property, we only need to figure out what to do for an implementation in a single place.

**Comment by berquist** [comment]
I'm thinking two stages:
- one PR to implement the fixes and add either unit or regression tests, not sure which is best, but leaves `homos` as-is
- another PR that handles the `homos <-> nelectrons` inversion
